lib/spack/spack/spec.py

```diff
@@ -3362,7 +3367,16 @@ def satisfies(self, other: Union[str, "Spec"], deps: bool = True) -> bool:
             return False
 
         if not self.variants.satisfies(other.variants):
-            return False
+            if all(not other.variants[k].propagate for k in other.variants):
+                return False
+
+            # Check if any not satisified variant is propagating and in spec's dependency
+            if any(
+                other.variants[variant].propagate
+                and not self.variant_exists_in_dependency(variant)
```

Comment:
@tgamblin This one couples the abstract spec semantic with packages, so the result varies depending on the repo:
```python
Spack version 0.23.0.dev0
Python 3.11.5, Linux x86_64
>>> import spack.spec
>>> s = spack.spec.Spec("hdf5++legacylaunchers")
>>> t = spack.spec.Spec("hdf5 ^openmpi")
>>> t.satisfies(s)
True
```
If we remove the variant from `openmpi` (or use another `mpi` provider), the same would be `False`. Are we fine with this semantic? Asking because up to know we tried hard not to couple things, so I'm wondering if this is deliberate, or accidental.

Reply:
@RikkiButler20 @tgamblin This one I believe is a bug:
```python
Spack version 0.23.0.dev0
Python 3.11.5, Linux x86_64
>>> import spack.spec
>>> s = spack.spec.Spec("hdf5++legacylaunchers")
>>> t = spack.spec.Spec("hdf5")
>>> t = spack.spec.Spec("hdf5 ^openmpi~legacylaunchers")
>>> t.satisfies(s)
True
```

Reply:
Here's another big logical inconsistency:
```python
Spack version 0.23.0.dev0
Python 3.11.5, Linux x86_64
>>> import spack.spec
>>> s = spack.spec.Spec("hdf5++legacylaunchers")
>>> t = spack.spec.Spec("hdf5 ^openmpi~legacylaunchers")
>>> u = spack.spec.Spec("hdf5+legacylaunchers")
>>> s == u
True
>>> t.satisfies(s)
True
>>> t.satisfies(u)
False
```
`s` and `u` are equal, but `t` satisfies only one of them. This in my opinion is kind of dangerous, since we might `dedupe` stuff in containers based on `==` and get different results in `satisfies` calls depending which spec got in the container first.

Reply:
There's also https://github.com/spack/spack/issues/47347 on `develop`, but I'll try to deal with that separately